### PR TITLE
Feat boosted pool add and remove

### PR DESCRIPTION
**Pull request by mkflow27**
Created a seperate issue #455 for tracking one outstanding todo.

### CODE REVIEW

File: src/entities/addLiquidityBoosted/doAddLiquidityPropotionalQuery.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import { BALANCER_COMPOSITE_LIQUIDITY_ROUTER, CHAINS } from '@/utils';
+
+import { AddLiquidityProportionalInput } from '../addLiquidity/types';
+import { Address } from '@/types';
+
+import { balancerCompositeLiquidityRouterAbi } from '@/abi';
+
+export const doAddLiquidityProportionalQuery = async (
+    { rpcUrl, chainId, referenceAmount }: AddLiquidityProportionalInput,
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/doAddLiquidityPropotionalQuery.ts:11]: I would change this to accept user address and userData. (Not as optionals, index.ts can handle defaults)

File: src/entities/addLiquidityBoosted/doAddLiquidityUnbalancedQuery.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import { BALANCER_COMPOSITE_LIQUIDITY_ROUTER, CHAINS } from '@/utils';
+
+import { AddLiquidityUnbalancedInput } from '../addLiquidity/types';
+import { Address } from '@/types';
+
+import { balancerCompositeLiquidityRouterAbi } from '@/abi';
+
+export const doAddLiquidityUnbalancedQuery = async (
+    { rpcUrl, chainId, amountsIn }: AddLiquidityUnbalancedInput,
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/doAddLiquidityUnbalancedQuery.ts:11]: Same as `doAddLiquidityProportionalQuery`

File: src/entities/addLiquidityBoosted/index.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import { balancerCompositeLiquidityRouterAbi, balancerRouterAbi } from '@/abi';
+
+import { Hex } from '@/types';
+
+export class AddLiquidityBoostedV3 {
+    async query(
+        input: AddLiquidityInput,
+        poolState: PoolState | PoolStateWithUnderlyings,
+    ): Promise<AddLiquidityBaseQueryOutput> {
+        // Technically possible to pass singleToken adds here due to type
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/index.ts:37]: Can change type to be 
`AddLiquidityUnbalancedInput
    | AddLiquidityProportionalInput;`

File: src/entities/addLiquidityBoosted/index.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        // disallow it for this class
+        // TODO: Use input validator
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/index.ts:39]: Couple of validations I can think of:
* Protocol version
* Input amounts are underlying

File: src/entities/addLiquidityBoosted/index.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                    input.referenceAmount.rawAmount,
+                );
+                break;
+            }
+            case AddLiquidityKind.SingleToken: {
+                throw new Error('SingleToken not supported');
+            }
+        }
+
+        const output: AddLiquidityBaseQueryOutput = {
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/index.ts:95]: In the nested PR I update to return userAddress and userData to buildCall can use that info.

File: src/entities/permit2Helper/index.ts [hunk 6]
```diff
@@ -114,6 +114,50 @@ export class Permit2Helper {
...
+        ) {
+            throw new Error(
+                "Expirations length doesn't match amountsIn length",
+            );
+        }
+        const amounts = getAmountsCall(input);
+        const spender = BALANCER_COMPOSITE_LIQUIDITY_ROUTER[input.chainId];
+        const details: PermitDetails[] = [];
+
+        // unbalanced:
```
Comment (johngrantuk) [inline src/entities/permit2Helper/index.ts:140]: Can remove all the debug stuff.

File: src/entities/removeLiquidityBoostedV3/doRemoveLiquidityProportionalQuery.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import { BALANCER_COMPOSITE_LIQUIDITY_ROUTER, CHAINS } from '@/utils';
+
+import { Address } from '@/types';
+
+import { balancerCompositeLiquidityRouterAbi } from '@/abi';
+
+import { RemoveLiquidityProportionalInput } from 'src';
+
+export const doRemoveLiquidityProportionalQuery = async (
+    { rpcUrl, chainId, bptIn }: RemoveLiquidityProportionalInput,
```
Comment (johngrantuk) [inline src/entities/removeLiquidityBoostedV3/doRemoveLiquidityProportionalQuery.ts:12]: similar to adds, can update to accept address and userData

File: src/entities/addLiquidityBoosted/index.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+import { getValue } from '@/entities/utils/getValue';
+import { encodeFunctionData } from 'viem';
+import { balancerCompositeLiquidityRouterAbi, balancerRouterAbi } from '@/abi';
+
+import { Hex } from '@/types';
+
+export class AddLiquidityBoostedV3 {
+    async query(
+        input: AddLiquidityInput,
+        poolState: PoolState | PoolStateWithUnderlyings,
```
Comment (johngrantuk) [inline src/entities/addLiquidityBoosted/index.ts:35]: Seems like there is something incorrect here? I don't think we can use `PoolState`?

File: src/entities/removeLiquidityBoostedV3/index.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        const amountsOutInNumbers = await doRemoveLiquidityProportionalQuery(
+            input,
+            poolState.address,
+        );
+
+        // amountsOut are in underlying Tokens.
+        const amountsOut = amountsOutInNumbers.map((amount, i) => {
+            const token = new Token(
+                input.chainId,
+                poolState.tokens[i].underlyingToken.address,
```
Comment (johngrantuk) [inline src/entities/removeLiquidityBoostedV3/index.ts:50]: This is one of your build issues.

File: test/v3/addLiquidityNested/addLiquidityNestedV3.integration.test.ts [hunk 10]
```diff
@@ -53,7 +53,7 @@ const usdcToken = new Token(chainId, USDC.address, USDC.decimals);
 const wethToken = new Token(chainId, WETH.address, WETH.decimals);
 const mainTokens = [wethToken, daiToken, usdcToken];
 
-describe.skip('V3 add liquidity nested test, with Permit2 direct approval', () => {
+describe('V3 add liquidity nested test, with Permit2 direct approval', () => {
```
Comment (johngrantuk) [inline test/v3/addLiquidityNested/addLiquidityNestedV3.integration.test.ts:56]: These skips are handled in my recent PR so you can just remove this diff to avoid conflicts.

File: test/v3/addLiquidityBoosted.integration.test.ts
Comment (johngrantuk) [inline test/v3/addLiquidityBoosted.integration.test.ts:1]: As discussed after your refactor ping me to let me know its good for review. I would recommend checking out the nested PR and maybe follow a similar approach which has two benefits imo: 1. Don't need to update all the existing test helpers which are convulted and hard to follow. 2. Makes the approval flow much easier to follow and confirm.

File: test/v3/removeLiquidityBoosted.integration.test.ts
Comment (johngrantuk) [inline test/v3/removeLiquidityBoosted.integration.test.ts:1]: Same as add tests.

File: test/v3/addLiquidityBoosted.integration.test.ts
Comment (mkflow27) [reply test/v3/addLiquidityBoosted.integration.test.ts:1]: I have refactored the tests to not use the helpers and instead do the intermediary steps as part of the tests. I did not change anything in the way how I handle approvals as I think the way they are done in the tests here are fine.

File: test/v3/removeLiquidityBoosted.integration.test.ts
Comment (mkflow27) [reply test/v3/removeLiquidityBoosted.integration.test.ts:1]: I utilise a helper here (which I think is ok) as a precondition for remove liq testing. Similarly I think the way I do approvals is ok. Compared to your tests, you do approvals right in the actualy tests, whereas I spread it.

File: src/entities/addLiquidityBoosted/doAddLiquidityPropotionalQuery.ts (hunk 1 above)
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/doAddLiquidityPropotionalQuery.ts:11]: I have addressed this as part of a larger refactor. It now allows to optionally supply `userData` & `userAddress` when querying the add/Remove classes and enforces values to be provided in the queries. Let me know what you think of that approach. It makes typing a bit overlapping as the same properties are either optionally or required.

File: src/entities/addLiquidityBoosted/doAddLiquidityUnbalancedQuery.ts (hunk 2 above)
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/doAddLiquidityUnbalancedQuery.ts:11]: addresses in a larger refactor to use userData and userAddress

File: src/entities/addLiquidityBoosted/index.ts (hunk 8 above)
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts:35]: The required state was actually dependend on the type. I now require to simply always pass a `PoolStateWithUnderlyings`. Only in the proportional adds was the type with underlying requires as the returned TokenAmounts need to have the pool's token underlyings but in the case of an unbalanced the returned TokenAmount is the BPT and all information was already shared in the input/poolstate.

File: src/entities/addLiquidityBoosted/index.ts [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                    return TokenAmount.fromRawAmount(
+                        new Token(input.chainId, t.address, t.decimals),
+                        t.rawAmount,
+                    );
+                });
+
+                break;
+            }
+            case AddLiquidityKind.Proportional: {
+                const exactAmountsInNumbers =
```
[the remark this answers is not in the document]
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts:63]: So the smart contracts will return the amounts of underlying tokens that the vault pulls in order to return an exactBpt out. So this variable just holds the numbers the SC returns. However the task is then below to return the TokenAmount as a combination of underlyingTokens & SC return values. I don't see how I would need to handle BPTs here. I need to handle the expected Tokens the Vault will pull in.

File: src/entities/permit2Helper/index.ts (hunk 6 above)
Comment (mkflow27) [reply src/entities/permit2Helper/index.ts:140]: addressed

File: src/entities/addLiquidityBoosted/index.ts [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        switch (input.addLiquidityKind) {
+            case AddLiquidityKind.Unbalanced: {
+                callData = encodeFunctionData({
+                    abi: balancerCompositeLiquidityRouterAbi,
+                    functionName: 'addLiquidityUnbalancedToERC4626Pool',
+                    args: [
+                        input.poolId,
+                        input.amountsIn.map((amount) => amount.amount),
+                        amounts.minimumBpt,
+                        !!input.wethIsEth,
```
[the remark this answers is not in the document]
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts:123]: addressed

File: src/entities/removeLiquidityBoostedV3/index.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        const amounts = getAmountsCall(input);
+
+        const callData = encodeFunctionData({
+            abi: balancerCompositeLiquidityRouterAbi,
+            functionName: 'removeLiquidityProportionalFromERC4626Pool',
+            args: [
+                input.poolId,
+                input.bptIn.amount,
+                amounts.minAmountsOut,
+                !!input.wethIsEth,
```
[the remark this answers is not in the document]
Comment (mkflow27) [reply src/entities/removeLiquidityBoostedV3/index.ts:84]: addressed

File: src/entities/removeLiquidityBoostedV3/index.ts (hunk 9 above)
Comment (mkflow27) [reply src/entities/removeLiquidityBoostedV3/index.ts]: build issues are addressed now

File: test/v3/addLiquidityNested/addLiquidityNestedV3.integration.test.ts (hunk 10 above)
Comment (mkflow27) [reply test/v3/addLiquidityNested/addLiquidityNestedV3.integration.test.ts:56]: removed the dif.

File: src/entities/removeLiquidityBoostedV3/doRemoveLiquidityProportionalQuery.ts (hunk 7 above)
Comment (mkflow27) [reply src/entities/removeLiquidityBoostedV3/doRemoveLiquidityProportionalQuery.ts]: Addressed

File: src/entities/addLiquidityBoosted/index.ts (hunk 5 above)
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts:102]: I have addressed this for add- as well as remove liquidity

File: src/entities/addLiquidityBoosted/index.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        // TODO: Use input validator
+        const bptToken = new Token(input.chainId, poolState.address, 18);
+
+        let bptOut: TokenAmount;
+        let amountsIn: TokenAmount[];
+        let tokenInIndex: number | undefined;
+
+        // Check if userAddress and userData were provided, and assign default values if not
+        if (!('userAddress' in input) || input.userAddress === undefined) {
+            // TODO: I think using a random address might be better than using the 0 address.
```
Comment (mkflow27) [inline src/entities/addLiquidityBoosted/index.ts:50]: @johngrantuk I think defaulting to the 0 address might not be the right choice but rather a random address. What do you think?

File: src/data/providers/balancer-api/modules/pool-state/index.ts [hunk 15]
```diff
@@ -39,6 +43,29 @@ export class Pools {
...
+          index
+          address
+          decimals
+          balance
+          underlyingToken {
+            address
+            decimals
+          }
+        }
+        dynamicData {
```
Comment (johngrantuk) [inline src/data/providers/balancer-api/modules/pool-state/index.ts:63]: I don't think we need this field?

File: src/entities/addLiquidityBoosted/index.ts (hunk 14 above)
Comment (johngrantuk) [reply src/entities/addLiquidityBoosted/index.ts:50]: I think zeroAddress is best but maybe you can explain why random is better?

File: src/entities/index.ts [hunk 16]
```diff
@@ -24,3 +24,5 @@ export * from './types';
 export * from './utils';
 export * from './swap';
 export * from './swap/swaps/v2/auraBalSwaps';
+export * from './addLiquidityBoosted';
+export * from './removeLiquidityBoostedV3';
```
Comment (johngrantuk) [inline src/entities/index.ts:28]: Noticed one has V3 and one doesn't, lets use a consistent pattern, I'd say without V3?

File: src/data/providers/balancer-api/modules/pool-state/index.ts (hunk 15 above)
Comment (mkflow27) [reply src/data/providers/balancer-api/modules/pool-state/index.ts:63]: You are right. I think I can actually remove the changes in the api, since I am not using the api in my tests, as we decided to use mock data. I'll remove it.

File: src/entities/addLiquidityBoosted/index.ts (hunk 14 above)
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts:50]: My reasoning was that it is more likely for a query to return different data than what is expected as it is with a random address. This is probably an exotic scenario, so overall unlikely. So for readability / potential debugging purposes address 0 makes it easiert to work with

File: src/entities/addLiquidityBoosted/index.ts [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+        // TODO: Use input validator
+        const bptToken = new Token(input.chainId, poolState.address, 18);
+
+        let bptOut: TokenAmount;
+        let amountsIn: TokenAmount[];
+        let tokenInIndex: number | undefined;
+        switch (input.kind) {
+            case AddLiquidityKind.Unbalanced: {
+                const bptAmountOut = await doAddLiquidityUnbalancedQuery(
+                    input,
```
[the remark this answers is not in the document]
Comment (mkflow27) [reply src/entities/addLiquidityBoosted/index.ts]: I have added this and now fill the amounts in default values in case the user did not provide all the tokenAmounts he wants to join with.

File: src/entities/index.ts (hunk 16 above)
Comment (mkflow27) [reply src/entities/index.ts]: Done. Renamed the folders to not have v3 anymore. For the classnames I found it easier to leave them with V3. Can also change. What do you think?

File: src/data/providers/balancer-api/modules/pool-state/index.ts [hunk 18]
```diff
@@ -76,4 +103,24 @@ export class Pools {
...
+            variables: {
+                id: id.toLowerCase(),
+                chain: API_CHAIN_NAMES[this.balancerApiClient.chainId],
+            },
+        });
+        const poolStateWithUnderlyings: PoolStateWithUnderlyings = {
+            ...data.poolGetPool,
+            tokens: data.poolGetPool.poolTokens,
+            type: mapPoolType(data.poolGetPool.type),
+            //underlyings: data.poolGetPool.poolTokens.map((token) => {token}),
```
[the remark this answers is not in the document]
Comment (mkflow27) [reply src/data/providers/balancer-api/modules/pool-state/index.ts:121]: removed.

File: src/entities/addLiquidityBoosted/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (johngrantuk) [reply src/entities/addLiquidityBoosted/index.ts:99]: Separate issue to follow up on this: https://github.com/balancer/b-sdk/issues/455
